File: #0402.remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {

        int size = num.size();
        if (k >= size) {
            return "0";
        }

        stack<int> st;
        for (int i = 0; i < size; i++) {
            while (k > 0 && !st.empty() && st.top() > num[i]) {
                st.pop();
                k--;
            }
            st.push(num[i]);
        }

        while (k-- > 0) {
            st.pop();
        }

        string result = "";
        while (!st.empty()) {
            result += st.top();
            st.pop();
        }

        std::reverse(result.begin(), result.end());
        while (result.size() > 1 && result[0] == '0') {
            result.erase(0, 1);
        }
        return result;
    }
};

File: #0402.remove-k-digits.js
/**
 * @param {string} num
 * @param {number} k
 * @return {string}
 */
var removeKdigits = function(num, k) {
    let stack = [];
    for (let i = 0; i < num.length; i++) {
        while (stack[stack.length - 1] > num[i] && k) {
            stack.pop();
            k--;
        }

        stack.push(num[i]);
        if (stack.length === 1 && stack[0] === "0") {
            stack.pop();
        }

        while (i === num.length - 1 && k) {
            stack.pop();
            k--;
        }
    }

    return stack.length !==0 ? stack.join("") : "0";
};
